### PR TITLE
Strip trailing slash when setting the password reset cookie

### LINKED ISSUE #1983: Password Key Invalid Issue
### Reproduction Steps

+ I think this is different from the WPEngine/Flywheel issue. 
+ HS-176845 (credentials in one pass)
+ Turn off all plugins but LifterLMS + LaunchPad
+ In an incognito window do a password reset

### Expected Behavior

+ Normally we see this with our resets the cookie stays even after the redirects:
image
image

### Actual Behavior

+ the user's site and staging site does not keep the cookie:
image

and the redirect seems to wipe it out:
image

### Error Messages / Logs

+ Include any relevant error messages or log files
```
<!-- Paste error logs / backtraces below this line -->

```

### System and Environment Information

<details>
<summary>System Report</summary>

```

Wordpress
-------------------------------------------

Home Url: [removed]
Site Url: [removed]
Login Url: [removed]/wp-login.php
Version: 5.9
Debug Mode: Yes
Debug Log: Yes
Debug Display: No
Locale: en_US
Multisite: No
Page For Posts: Course Catalog (#80) [[removed]/course-catalog]
Page On Front: Home page (#4) [[removed]/]
Permalink Structure: /'/%year%/%monthnum%/%day%/%postname%/'
Show On Front: page
Wp Cron: Yes


Settings
-------------------------------------------

Version: 5.8.0
Db Version: 5.8.0
Course Catalog: Course Catalog (#80) [[removed]/course-catalog]
Membership Catalog: Membership (#82) [[removed]/membership]
Student Dashboard: Account Dashboard (#86) [[removed]/account-dashboard]
Checkout Page: Checkout (#88) [[removed]/checkout]
Course Catalog Per Page: 12
Course Catalog Sorting: date,DESC
Membership Catalog Per Page: 10
Membership Catalog Sorting: title,ASC
Site Membership: Not Set
Courses Endpoint: my-courses
Edit Endpoint: edit-account
Lost Password Endpoint: lost-password
Vouchers Endpoint: redeem-voucher
Autogenerate Username: no
Password Strength Meter: no
Minimum Password Strength: weak
Terms Required: no
Terms Page: Terms and Conditions (#84) [[removed]/terms-and-conditions]
Checkout Names: required
Checkout Address: required
Checkout Phone: required
Checkout Email Confirmation: yes
Open Registration: yes
Registration Names: required
Registration Address: required
Registration Phone: required
Registration Voucher: hidden
Registration Email Confirmation: yes
Account Names: required
Account Address: required
Account Phone: required
Account Email Confirmation: yes
Confirmation Endpoint: confirm-payment
Force Ssl Checkout: yes
Country: US
Currency: USD
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Decimals: 2
Trim Zero Decimals: no
Recurring Payments: no
Email From Address: [removed]
Email From Name: [removed]
Email Footer Text: Got questions? Hit reply. We're here for you.
Email Header Image: [removed]/wp-content/uploads/2017/02/email_header-e1486031128828.gif
Cert Bg Width: 800
Cert Bg Height: 616
Cert Legacy Compat: no


Constants
-------------------------------------------

LLMS_REMOVE_ALL_DATA: undefined
LLMS_REST_DISABLE: undefined
LLMS_SITE_FEATURE_RECURRING_PAYMENTS: undefined
LLMS_SITE_IS_CLONE: undefined


Gateways
-------------------------------------------

Manual: Enabled
Manual Logging: no
Manual Order: 3


Server
-------------------------------------------

Mysql Version: 5.7.23
Php Curl: Yes
Php Default Timezone: UTC
Php Fsockopen: Yes
Php Max Input Vars: 1000
Php Max Upload Size: 512 MB
Php Memory Limit: 256M
Php Post Max Size: 516M
Php Soap: Yes
Php Suhosin: No
Php Time Limt: 30
Php Version: 7.3.33
Software: Apache
Wp Memory Limit: 40M


Browser
-------------------------------------------

HTTP USER AGENT: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.80 Safari/537.36


Theme
-------------------------------------------

Name: LifterLMS LaunchPad
Version: 2.4.9
Themeuri: https://lifterlms.com/launchpad/
Authoruri: https://gocodebox.com
Template: 
Child Theme: No
Llms Support: No


Plugins
-------------------------------------------

LifterLMS: 5.8.0


Integrations
-------------------------------------------

BbPress: No
BuddyPress: No


Template Overrides
-------------------------------------------



```

</details>

This issue has be recreated:
+ [ ] Locally
+ [x] On a staging site
+ [x] On a production website
+ [x] With only LifterLMS and a default theme

### Browser, Device, and Operating System Information

+ Browser name and version
+ Operating System name and version
+ Device name and version (if applicable)

### PR DESCRIPTION
**Pull request by thomasplevy**
## Description

Fixes #1983

## How has this been tested?

+ Manually
+ New & updated phpunit tests

## Types of changes
 Bugfix

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
This doesn't work for me.
I mean it works if my permalink structure is `/%postname%` as you said but then doesn't work if my permalink structure is (as it _should_) `/%postname%/` for the same reason, the cookie path is set to have not trailing slash but I land on an URL with a trailing slash.
I propose:
a) we fix the way we generate the endpoint URLs, adding a trailing slash ONLY if `get_option( 'permalink_structure' )` ends with a slash.
OR
b) we strip the trailing slash, as you suggest, but ONLY if the `get_option( 'permalink_structure' )` doesn't end with a slash.

**Comment by thomasplevy** [comment]
Oh shoot you're right... I tested this like an idiot and didn't clear my cookies between changing the permalinks so it worked the second time because the cookie was still set from the other version

**Comment by thomasplevy** [comment]
I tried real hard to avoid rewriting the endpoint function... I'll dive into that now I guess

**Comment by eri-trabiccolo** [comment]
> I tried real hard to avoid rewriting the endpoint function... I'll dive into that now I guess

My idea is simple. Take this:
https://github.com/gocodebox/lifterlms/blob/5.8.0/includes/functions/llms.functions.page.php#L94

we have to, check whether or not `get_option( 'permalink_structure' );` ends with a `/`. Then BEFORE adding the query string, if that doesn't end with a /, `untrailslashit`, then add the query.

**Comment by thomasplevy** [comment]
mmm yea... look what I found: https://developer.wordpress.org/reference/functions/user_trailingslashit/

**Comment by eri-trabiccolo** [comment]
Jeez you're a goldminer! <3

**Comment by thomasplevy** [comment]
Alright this works much better.

File: includes/functions/llms.functions.page.php [hunk 1]
```diff
@@ -62,37 +62,44 @@ function llms_confirm_payment_url( $order_key = null ) {
...
+ * @param string $value     Endpoint query parameter value.
  * @param string $permalink Base URL to append the endpoint to. Optional, uses the current page when not supplied.
  * @return string
  */
 function llms_get_endpoint_url( $endpoint, $value = '', $permalink = '' ) {
 
-	if ( ! $permalink ) {
-		$permalink = get_permalink();
-	}
+	$permalink = $permalink ? $permalink : get_permalink();
```
Comment (pondermatic) [inline includes/functions/llms.functions.page.php:79]: Short [ternary operator](https://www.php.net/manual/en/language.operators.comparison.php#language.operators.comparison.ternary).
```suggestion
	$permalink = $permalink ?: get_permalink();
```

File: includes/functions/llms.functions.page.php [hunk 2]
```diff
@@ -62,37 +62,44 @@ function llms_confirm_payment_url( $order_key = null ) {
...
 
 	// Map endpoint to options.
 	$vars     = LLMS()->query->get_query_vars();
 	$endpoint = isset( $vars[ $endpoint ] ) ? $vars[ $endpoint ] : $endpoint;
```
Comment (pondermatic) [inline includes/functions/llms.functions.page.php]: [Null coalescing operator](https://www.php.net/manual/en/language.operators.comparison.php#language.operators.comparison.coalesce).
```suggestion
	$endpoint = $vars[ $endpoint ] ?? $endpoint;
```

File: tests/phpunit/unit-tests/functions/class-llms-test-functions-page.php [hunk 3]
```diff
@@ -41,6 +41,108 @@ public function test_llms_confirm_payment_url() {
...
+			$this->assertEquals( "{$permalink}{$slug}/test/", llms_get_endpoint_url( $var, 'test' ) );
+			$this->assertEquals( "https://fake.tld/{$slug}/1/", llms_get_endpoint_url( $var, 1, 'https://fake.tld/' ) );
+			$this->assertEquals( "https://fake.tld/{$slug}/1/?whatever=yes", llms_get_endpoint_url( $var, 1, 'https://fake.tld/?whatever=yes' ) );
+
+		}
+
+		$this->go_to( '' );
+
+		update_option( 'permalink_structure', $orig_permastruct );
+		global $wp_rewrite;
```
Comment (pondermatic) [inline tests/phpunit/unit-tests/functions/class-llms-test-functions-page.php]: Traditionally in PHP, use of the global keyword happens at the start of the function.

File: tests/phpunit/unit-tests/functions/class-llms-test-functions-page.php [hunk 4]
```diff
@@ -41,6 +41,108 @@ public function test_llms_confirm_payment_url() {
...
+			$this->assertEquals( "{$permalink}/{$slug}/test", llms_get_endpoint_url( $var, 'test' ) );
+			$this->assertEquals( "https://fake.tld/{$slug}/1", llms_get_endpoint_url( $var, 1, 'https://fake.tld/' ) );
+			$this->assertEquals( "https://fake.tld/{$slug}/1?whatever=yes", llms_get_endpoint_url( $var, 1, 'https://fake.tld/?whatever=yes' ) );
+
+		}
+
+		$this->go_to( '' );
+
+		update_option( 'permalink_structure', $orig_permastruct );
+		global $wp_rewrite;
```
Comment (pondermatic) [inline tests/phpunit/unit-tests/functions/class-llms-test-functions-page.php]: Traditionally in PHP, use of the global keyword happens at the start of the function.

File: tests/phpunit/unit-tests/functions/class-llms-test-functions-page.php [hunk 5]
```diff
@@ -41,6 +41,108 @@ public function test_llms_confirm_payment_url() {
...
+	 * @return void
+	 */
+	public function test_llms_get_endpoint_url_without_trailing_slash_pretty_permalink() {
+
+		$orig_permastruct = get_option( 'permalink_structure' );
+
+		LLMS_Install::create_pages();
+
+		update_option( 'permalink_structure', '/%postname%' );
+		global $wp_rewrite;
```
Comment (pondermatic) [inline tests/phpunit/unit-tests/functions/class-llms-test-functions-page.php]: Traditionally in PHP, use of the global keyword happens at the start of the function.

File: tests/phpunit/unit-tests/functions/class-llms-test-functions-page.php [hunk 6]
```diff
@@ -41,6 +41,108 @@ public function test_llms_confirm_payment_url() {
...
+	 * @return void
+	 */
+	public function test_llms_get_endpoint_url_with_trailing_slash_pretty_permalink() {
+
+		$orig_permastruct = get_option( 'permalink_structure' );
+
+		LLMS_Install::create_pages();
+
+		update_option( 'permalink_structure', '/%postname%/' );
+		global $wp_rewrite;
```
Comment (pondermatic) [inline tests/phpunit/unit-tests/functions/class-llms-test-functions-page.php]: Traditionally in PHP, use of the global keyword happens at the start of the function.

**Review by pondermatic — COMMENTED** [review]
The position of the `global` keyword is no big deal. Might want to only delcare `global $wp_rewrite;` once per function though.

File: includes/functions/llms.functions.page.php (hunk 1 above)
Comment (thomasplevy) [reply includes/functions/llms.functions.page.php:79]: WP forbids the use of short ternary so we do as well: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#ternary-operator

The null coalesce *is allowed*, however.

Comment (pondermatic) [reply includes/functions/llms.functions.page.php:79]: Interesting! Thanks for the reference. I had forgotten that one.

**Review by pondermatic — APPROVED** [review]

**Pull request merged**